### PR TITLE
Type the operand of code_declt

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f537e7cdf73b7564971082f2f4a81fe6947d3564).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90172916
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/cpp/cpp_util.cpp [hunk 1]
```diff
@@ -11,12 +11,12 @@
...
-exprt cpp_symbol_expr(const symbolt &symbol)
+symbol_exprt cpp_symbol_expr(const symbolt &symbol)
 {
   symbol_exprt tmp(symbol.name, symbol.type);
 
   if(symbol.is_lvalue)
     tmp.set(ID_C_lvalue, true);
 
-  return std::move(tmp);
+  return tmp;
```
Comment (tautschnig) [inline src/cpp/cpp_util.cpp:21]: Why is the `std::move` being removed here?

File: src/goto-instrument/dump_c.cpp [hunk 2]
```diff
@@ -894,7 +894,9 @@ void dump_ct::convert_global_variable(
   }
 
   if(!func.empty() && !symbol.is_extern)
-    local_static_decls[symbol.name]=d;
+  {
+    local_static_decls.insert(std::pair<irep_idt, code_declt>(symbol.name, d));
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: I _think_ that's fine, but I am surprised to see this seemingly unrelated change in here. And then `.emplace` would seem better?!

File: src/cpp/cpp_util.cpp (hunk 1 above)
Comment (kroening) [reply src/cpp/cpp_util.cpp:21]: Perhaps counterintuitively, the std::move _prevents_ a move here. I.e., removing it enables copy elision.

File: src/goto-instrument/dump_c.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-instrument/dump_c.cpp]: It's related as follows: the PR enforces that a code_declt always has a symbol_typet parameter, which is not the case for a very brief time when using [] to insert an element into a map.
But will use .emplace.

File: src/cpp/cpp_util.cpp (hunk 1 above)
Comment (tautschnig) [reply src/cpp/cpp_util.cpp:21]: Why would that be the case? Notably, clang-7 had complained about the absence of `std::move` here, and it has just been introduced in 72bfc352035.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 35b893c9b4d0dfbdaf6b023e6933b01fae0cec64).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90241109
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/cpp/cpp_util.cpp (hunk 1 above)
Comment (smowton) [reply src/cpp/cpp_util.cpp:21]: @tautschnig the key is whether a type conversion is required or not. Previously we returned a `symbol_exprt`, meaning we needed a conversion to `exprt` (via its copy- or move-constructor), so `std::move` was a good thing (picking the move constructor). Now we return a `symbol_exprt`, meaning no type conversion is necessary and RVO / copy/move-elision is possible, so `std::move` is now a *bad* thing.

**Review by smowton — APPROVED** [review]

File: src/cpp/cpp_util.cpp (hunk 1 above)
Comment (tautschnig) [reply src/cpp/cpp_util.cpp:21]: Many thanks for the elaboration!

**Review by tautschnig — APPROVED** [review]

**Review by danpoe — APPROVED** [review]

**Review by martin-cs — APPROVED** [review]
My part looks fine.

**Comment by tautschnig** [comment]
Needs a rebase.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e889df7effb03c33b084c883797d559da916b0fd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90402530
